### PR TITLE
rafthttp: extra resp.Body.Close paranoia

### PR DESCRIPTION
**Pull request by heyitsanthony**
Be more paranoid about closing a response body.

Pursuant to #6933

### CODE REVIEW

**Comment by gyuho** [comment]
LGTM.

File: rafthttp/pipeline.go [hunk 1]
```diff
@@ -124,7 +132,6 @@ func (p *pipeline) post(data []byte) (err error) {
 		select {
 		case <-done:
 		case <-p.stopc:
-			waitSchedule()
```
Comment (xiang90) [inline rafthttp/pipeline.go]: do you know why we have this before?

Comment (heyitsanthony) [inline rafthttp/pipeline.go]: Hold over from before ctxs:
```
    rafthttp: wait 1ms before enabling cancel
    
    CancelRequest only effects on in-flight request, so we need to wait
    for Do(request) called before enabling cancel.
```

**Comment by xiang90** [comment]
LGTM

**Comment by heyitsanthony** [comment]
still leaking http resources on ci, unfortunately

File: rafthttp/pipeline.go [hunk 2]
```diff
@@ -136,11 +143,11 @@ func (p *pipeline) post(data []byte) (err error) {
 		return err
 	}
 	b, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
```
Comment (upccup) [inline rafthttp/pipeline.go]: a little confuse why not use defer resp.Body.Close()

Comment (heyitsanthony) [inline rafthttp/pipeline.go]: defer doesn't buy anything here; the function won't panic so it'll only be needless overhead.

File: rafthttp/snapshot_sender.go [hunk 3]
```diff
@@ -126,8 +126,17 @@ func (s *snapshotSender) post(req *http.Request) (err error) {
...
+		tc := make(chan struct{})
+		t := time.AfterFunc(snapResponseReadTimeout, func() {
+			defer close(tc)
+			httputil.GracefulClose(resp)
+		})
 		body, err := ioutil.ReadAll(resp.Body)
+		if t.Stop() {
+			resp.Body.Close()
+		} else {
+			<-tc
```
Comment (odeke-em) [inline rafthttp/snapshot_sender.go]: resp.Body won't be closed in this case, is this intended?  I am not familiar with all the code but this looks like one of the mention elusive leaks.

File: rafthttp/stream.go [hunk 4]
```diff
@@ -469,11 +470,11 @@ func (cr *streamReader) dial(t streamType) (io.ReadCloser, error) {
 		return nil, fmt.Errorf("peer %s failed to find local node %s", cr.peerID, cr.tr.ID)
 	case http.StatusPreconditionFailed:
 		b, err := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
```
Comment (odeke-em) [inline rafthttp/stream.go]: Why not move the httputil.GracefulClose from below that you removed, up here?

**Pull request closed without merging**